Mathlib/Analysis/Meromorphic/NormalFormAt.lean

```diff
@@ -0,0 +1,240 @@
+    · exact Or.inr ⟨n, h, h₁h, h₂h, hfg.symm.trans h₃h⟩
+  · rintro (h | ⟨n, h, h₁h, h₂h, h₃h⟩)
+    · exact Or.inl (hfg.trans h)
+    · exact Or.inr ⟨n, h, h₁h, h₂h, hfg.trans h₃h⟩
+
+/-!
+## Relation to other properties of functions
+-/
+
+/-- If a function is meromorphic in normal form at `x`, then it is meromorphic at `x`. -/
+theorem MeromorphicNFAt.meromorphicAt (hf : MeromorphicNFAt f x) :
+    MeromorphicAt f x := by
+  rw [meromorphicNFAt_iff_analyticAt_or] at hf
+  rcases hf with h | h
+  · exact h.meromorphicAt
+  · obtain ⟨hf, _⟩ := h
+    exact hf
+
+/-- If a function is meromorphic in normal form at `x`, then it has non-negative order iff it is
+analytic. -/
+theorem MeromorphicNFAt.nonneg_order_iff_analyticAt (hf : MeromorphicNFAt f x) :
+    0 ≤ hf.meromorphicAt.order ↔ AnalyticAt 𝕜 f x := by
+  constructor <;> intro h₂f
+  · rw [meromorphicNFAt_iff_analyticAt_or] at hf
+    rcases hf with h | ⟨_, h₃f, _⟩
+    · exact h
+    · by_contra h'
+      exact lt_irrefl 0 (lt_of_le_of_lt h₂f h₃f)
+  · rw [h₂f.meromorphicAt_order]
+    simp
+
+/-- Analytic functions are meromorphic in normal form. -/
+theorem AnalyticAt.MeromorphicNFAt (hf : AnalyticAt 𝕜 f x) :
+    MeromorphicNFAt f x := by
+  simp [meromorphicNFAt_iff_analyticAt_or, hf]
+
+/-!
+## Continuous extension and conversion to normal form
+-/
+
+variable (f) (x) in
+/-- If `f` is meromorphic at `x`, convert `f` to normal form at `x` by changing its value at `x`.
+Otherwise, returns the 0 function. -/
+noncomputable def toMeromorphicNFAt :
+    𝕜 → E := by
+  by_cases hf : MeromorphicAt f x
+  · classical -- do not complain about decidability issues in Function.update
+    apply Function.update f x
+    by_cases h₁f : hf.order = (0 : ℤ)
+    · rw [hf.order_eq_int_iff] at h₁f
+      exact (Classical.choose h₁f) x
+    · exact 0
+  · exact 0
+
+/-- Conversion to normal form at `x` by changes the value only at x. -/
+lemma MeromorphicAt.eqOn_compl_singleton_toNF (hf : MeromorphicAt f x) :
+    Set.EqOn f (toMeromorphicNFAt f x) {x}ᶜ :=
+  fun _ _ ↦ by simp_all [toMeromorphicNFAt]
+
+/-- Conversion to normal form at `x` changes the value only at x. -/
+lemma MeromorphicAt.toNF_id_on_nhdNE (hf : MeromorphicAt f x) :
+    f =ᶠ[𝓝[≠] x] toMeromorphicNFAt f x :=
+  eventually_nhdsWithin_of_forall (fun _ hz ↦ hf.eqOn_compl_singleton_toNF hz)
+
+-- Two analytic functions agree on a punctured neighborhood iff they agree on a neighborhood.
+private lemma AnalyticAt.eventuallyEq_nhdNE_iff_eventuallyEq_nhd {g : 𝕜 → E} {z₀ : 𝕜}
+  (hf : AnalyticAt 𝕜 f z₀) (hg : AnalyticAt 𝕜 g z₀) (hfg : f =ᶠ[𝓝[≠] z₀] g) :
+    f =ᶠ[𝓝 z₀] g := by
+  rcases ((hf.sub hg).eventually_eq_zero_or_eventually_ne_zero) with h | h
+  · exact Filter.eventuallyEq_iff_sub.2 h
+  · simpa using (Filter.eventually_and.2 ⟨Filter.eventuallyEq_iff_sub.mp hfg, h⟩).exists
+
+/-- After conversion to normal form at `x`, the function has normal form. -/
+theorem MeromorphicAt.meromorphicNFAt_toNF (hf : MeromorphicAt f x) :
```

Comment:
```suggestion
protected theorem MeromorphicNFAt.toMeromorphicNFAt :
```
(if `hf` doesn't hold, then `toMeromorphicNFAt f x = 0` is in mero-nf)

Reply:
I accept you suggestion, but I am not 100% sure what it does. What does the `protected` do?